src/ILCompiler.Compiler/src/Compiler/HardwareIntrinsicHelpers.cs

```diff
@@ -40,6 +66,134 @@ public static bool IsHardwareIntrinsic(MethodDesc method)
             return false;
         }
 
+        /// <summary>
+        /// Cache used to speed up resolution of types referring to HW-dependent types.
+        /// Please note this is not "just a cache", we also use it to break circles in the
+        /// recursive type graph (e.g. int implements several generic interfaces instantiated over int).
+        /// </summary>
+        private static Dictionary<TypeDesc, bool> _hardwareIntrinsicDependentTypeCache = new Dictionary<TypeDesc, bool>();
+
+        /// <summary>
+        /// Gets a value indicating whether a given type doesn't refer to any HW intrinsics and can be safely
+        /// pre-JITted. This requires a recursive scan roughly mimicking the CoreCLR MethodTableBuilder behavior;
+        /// in particular, we check the following aspects:
+        /// (*) The type itself;
+        /// (*) For DefTypes, we recursively check all fields on the type;
+        /// (*) For generic types, we recursively check all instantiation parameters;
+        /// (*) For parameterized types, we recursively check the parameter type.
+        /// </summary>
+        /// <param name="type">Type to analyze</param>
+        /// <returns>True when a given type is the owner type for a hardware intrinsic</returns>
+        public static bool IsHardwareIntrinsicDependentType(TypeDesc type)
+        {
+            if (_hardwareIntrinsicDependentTypeCache.TryGetValue(type, out bool result))
+            {
+                return result;
+            }
+
+            // Initially mark the type in the cache as HW-independent so that, if we hit it
```

Comment:
When we make the compiler multi-threaded this will create a subtle bug since there may be multiple concurrent lookups.  Since this is basically DFS on the type graph, we could pass the visited set for the current expansion around so each thread's recursive descent is separate and only the final result would need synchronizing when we write to `_hardwareIntrinsicDependentTypeCache`.

Reply:
We don't use statics for anything that is mutable, not thread safe, or keeps a reference to a `TypeSystemContext` or `NodeFactory`. We would like to keep the ability to have multiple type system contexts in the same process. Hanging onto type system entities from statics makes that problematic.

I would prefer if we can solve this by giving these types indeterminate layout (no locking needed and the perf would be better).

Reply:
(The main reason for avoiding accumulating stuff in statics is because a lot of the compiler is also used in the JIT codebase that recycles type system contexts and any static caches like this are just memory leaks.)